move-vm-backend/src/abi.rs

```diff
@@ -0,0 +1,244 @@
+                .type_parameters
+                .iter()
+                .map(|tp| TypeAbilities::from(&tp.constraints))
+                .collect();
+
+            let fields = match &sdef.field_information {
+                StructFieldInformation::Native => vec![],
+                StructFieldInformation::Declared(defs) => defs
+                    .iter()
+                    .map(|field| Field {
+                        name: module.identifier_at(field.name).to_owned(),
+                        tp: make_type(&field.signature.0, module),
+                    })
+                    .collect(),
+            };
+
+            Struct {
+                name: module.identifier_at(handle.name).to_owned(),
+                type_parameters,
+                abilities: TypeAbilities::from(&handle.abilities),
+                fields,
+            }
+        })
+        .collect()
+}
+
+fn make_type(tok: &SignatureToken, module: &CompiledModule) -> Type {
+    match tok {
+        SignatureToken::Bool => Type::Bool,
+        SignatureToken::U8 => Type::U8,
+        SignatureToken::U16 => Type::U16,
+        SignatureToken::U32 => Type::U32,
+        SignatureToken::U64 => Type::U64,
+        SignatureToken::U128 => Type::U128,
+        SignatureToken::U256 => Type::U256,
+        SignatureToken::Address => Type::Address,
+        SignatureToken::Signer => Type::Signer,
+        SignatureToken::Vector(tp) => Type::Vector(Box::new(make_type(tp, module))),
+        SignatureToken::Struct(idx) => Type::Struct(make_struct_def(*idx, &[], module)),
+        SignatureToken::StructInstantiation(idx, tps) => {
+            Type::Struct(make_struct_def(*idx, tps, module))
+        }
+        SignatureToken::Reference(rf) => Type::Reference(Box::new(make_type(rf, module))),
+        SignatureToken::MutableReference(tp) => {
+            Type::MutableReference(Box::new(make_type(tp, module)))
+        }
+        SignatureToken::TypeParameter(val) => Type::TypeParameter(*val),
+    }
+}
+
+fn make_struct_def(
+    idx: StructHandleIndex,
+    tps: &[SignatureToken],
+    module: &CompiledModule,
+) -> StructDef {
+    let struct_handle = module.struct_handle_at(idx);
+    let struct_module_handle = module.module_handle_at(struct_handle.module);
+    let id = module.module_id_for_handle(struct_module_handle);
+
+    StructDef {
+        id,
+        name: module.identifier_at(struct_handle.name).to_owned(),
+        type_parameters: tps.iter().map(|tok| make_type(tok, module)).collect(),
+    }
+}
+
+fn make_func_abi(module: &CompiledModule) -> Vec<Func> {
+    module
+        .function_defs()
+        .iter()
+        .filter(|def| match def.visibility {
+            Visibility::Public | Visibility::Friend => true,
+            Visibility::Private => false,
+        })
+        .map(|def| {
+            let handle = module.function_handle_at(def.function);
+            let parameters = &module.signature_at(handle.parameters).0;
```

Comment:
Just a thought - it's hard to read it without any comment what `.0` means when not importing that into any IDE. 
GH review panel can't tell it (yet) 😸

Reply:
Done